Zotero/Controllers/IdentifierLookupController.swift

```diff
@@ -0,0 +1,445 @@
+        accessQueue.async(flags: .barrier) { [weak self] in
+            var lookupData: [LookupData]?
+            defer {
+                completion(lookupData)
+            }
+            guard let self = self else { return }
+            let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+            if self.lookupWebViewHandlersByLookupSettings[lookupSettings] != nil {
+                lookupData = self.lookupData
+                return
+            }
+            inMainThread(sync: true) {
+                if let webView = self.webViewProvider?.addWebView() {
+                    let lookupWebViewHandler = LookupWebViewHandler(lookupSettings: lookupSettings, webView: webView, translatorsController: self.translatorsController)
+                    self.lookupWebViewHandlersByLookupSettings[lookupSettings] = lookupWebViewHandler
+                }
+            }
+            guard let lookupWebViewHandler = self.lookupWebViewHandlersByLookupSettings[lookupSettings] else {
+                DDLogError("IdentifierLookupController: can't create LookupWebViewHandler instance")
+                return
+            }
+            self.setupObserver(for: lookupWebViewHandler)
+            lookupData = self.lookupData
+        }
+    }
+    
+    func lookUp(libraryId: LibraryIdentifier, collectionKeys: Set<String>, identifier: String) {
+        let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+        lookupWebViewHandlersByLookupSettings[lookupSettings]?.lookUp(identifier: identifier)
+    }
+    
+    func cancelAllLookups() {
+        accessQueue.async(flags: .barrier) { [weak self] in
+            guard let self else { return }
+            DDLogInfo("IdentifierLookupController: cancel all lookups")
+            let keys = self.lookupWebViewHandlersByLookupSettings.keys
+            for key in keys {
+                if let webView = self.lookupWebViewHandlersByLookupSettings.removeValue(forKey: key)?.webViewHandler.webView {
+                    inMainThread {
+                        self.webViewProvider?.removeWebView(webView)
+                    }
+                }
+            }
+            self.remoteFileDownloader.stop()
+            let lookupData = self.lookupData
+            cleanupLookup(force: true, alreadyInQueue: true, update: Update(kind: .finishedAllLookups))
+            let storedItemResponses: [(ItemResponse, LibraryIdentifier)] = lookupData.compactMap {
+                switch $0.state {
+                case .translated(let translatedLookupData):
+                    return (translatedLookupData.response, translatedLookupData.libraryId)
+                    
+                default:
+                    return nil
+                }
+            }
+            for (response, libraryId) in storedItemResponses {
```

Comment:
Call `self.backgroundQueue.async` before the `for` loop, unnecessary overhead. You can also create all dbrequests in the `for` loop (or `map`) and then call single `dbStorage.perform` with all requests for single db write.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/f73cdf75e9b42a27d48d096deeff37488fd0989d